File: package/ios/MarkerViewportFilter.swift
import MapKit

/// Selects a zoom-appropriate marker subset for large datasets.
enum MarkerViewportFilter {
  /// Precise viewport filter + spatial subsample over pre-narrowed candidates.
  ///
  /// The caller (spatial index) has already restricted `candidates` to cells
  /// near the region, so this runs over a small set and is safe to call off the
  /// main thread.
  static func displaySubset(
    candidates: [MarkerDescriptor],
    region: MKCoordinateRegion
  ) -> [MarkerDescriptor] {
    let maxCount = maxMarkers(for: region.span.latitudeDelta)
    let visible = candidates.filter { region.contains($0.coordinate, padding: 0.2) }

    guard visible.count > maxCount else {
      return visible
    }

    return spatialSubsample(visible, maxCount: maxCount, region: region)
  }

  static func markersFingerprint(_ markers: [MarkerDescriptor]?) -> Int {
    guard let markers, !markers.isEmpty else {
      return 0
    }

    var hasher = Hasher()
    hasher.combine(markers.count)
    for marker in markers {
      hasher.combine(marker.id)
      hasher.combine(marker.coordinate.latitude)
      hasher.combine(marker.coordinate.longitude)
      hasher.combine(marker.title)
      hasher.combine(marker.subtitle)
      hasher.combine(marker.draggable)
      hasher.combine(marker.clusterable)
    }
    return hasher.finalize()
  }

  /// Picks at most one marker per geographic cell so subsampling stays visually even.
  private static func spatialSubsample(
    _ markers: [MarkerDescriptor],
    maxCount: Int,
    region: MKCoordinateRegion
  ) -> [MarkerDescriptor] {
    let columns = Int(ceil(sqrt(Double(maxCount))))
    let rows = Int(ceil(Double(maxCount) / Double(columns)))

    let latMin = region.center.latitude - region.span.latitudeDelta * 0.6
    let latMax = region.center.latitude + region.span.latitudeDelta * 0.6
    let lonMin = region.center.longitude - region.span.longitudeDelta * 0.6
    let lonMax = region.center.longitude + region.span.longitudeDelta * 0.6

    let latStep = max(1e-9, (latMax - latMin) / Double(rows))
    let lonStep = max(1e-9, (lonMax - lonMin) / Double(columns))

    var buckets: [String: [MarkerDescriptor]] = [:]
    buckets.reserveCapacity(maxCount)

    for marker in markers {
      let row = min(rows - 1, max(0, Int((marker.coordinate.latitude - latMin) / latStep)))
      let column = min(columns - 1, max(0, Int((marker.coordinate.longitude - lonMin) / lonStep)))
      buckets["\(row)-\(column)", default: []].append(marker)
    }

    return buckets.values.map { cell in
      cell[cell.count / 2]
    }
  }

  private static func maxMarkers(for latitudeDelta: Double) -> Int {
    if latitudeDelta < 0.08 {
      return 2_000
    }
    if latitudeDelta < 0.5 {
      return 800
    }
    if latitudeDelta < 2.0 {
      return 350
    }
    return 200
  }
}

private extension MKCoordinateRegion {
  func contains(_ coordinate: Coordinate, padding: Double) -> Bool {
    let latPadding = span.latitudeDelta * padding
    let lonPadding = span.longitudeDelta * padding
    let minLat = center.latitude - span.latitudeDelta / 2 - latPadding
    let maxLat = center.latitude + span.latitudeDelta / 2 + latPadding
    let minLon = center.longitude - span.longitudeDelta / 2 - lonPadding
    let maxLon = center.longitude + span.longitudeDelta / 2 + lonPadding

    return coordinate.latitude >= minLat
      && coordinate.latitude <= maxLat
      && coordinate.longitude >= minLon
      && coordinate.longitude <= maxLon
  }
}

File: package/android/src/main/java/com/margelo/nitro/nitromaps/MarkerViewportFilter.kt
package com.margelo.nitro.nitromaps

import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.sqrt

internal object MarkerViewportFilter {
  /**
   * Precise viewport filter + spatial subsample over pre-narrowed candidates.
   *
   * The caller (spatial index) has already restricted [candidates] to cells
   * near the bounds, so this runs over a small set and is safe to call off the
   * main thread.
   */
  fun displaySubset(
    candidates: List<MarkerDescriptor>,
    bounds: LatLngBounds,
    latitudeSpan: Double,
  ): List<MarkerDescriptor> {
    val maxCount = maxMarkersForZoom(latitudeSpan)
    val paddedBounds = bounds.expandBy(0.2)

    val visible = candidates.filter { descriptor ->
      paddedBounds.contains(
        LatLng(descriptor.coordinate.latitude, descriptor.coordinate.longitude),
      )
    }

    if (visible.size <= maxCount) {
      return visible
    }

    return spatialSubsample(visible, maxCount, bounds).toList()
  }

  fun markersFingerprint(descriptors: Array<MarkerDescriptor>?): Int {
    if (descriptors.isNullOrEmpty()) {
      return 0
    }

    var hash = descriptors.size
    for (descriptor in descriptors) {
      hash = 31 * hash + descriptor.id.hashCode()
      hash = 31 * hash + descriptor.coordinate.latitude.hashCode()
      hash = 31 * hash + descriptor.coordinate.longitude.hashCode()
      hash = 31 * hash + (descriptor.title?.hashCode() ?: 0)
      hash = 31 * hash + (descriptor.subtitle?.hashCode() ?: 0)
      hash = 31 * hash + (descriptor.draggable?.hashCode() ?: 0)
      hash = 31 * hash + (descriptor.clusterable?.hashCode() ?: 0)
      hash = 31 * hash + (descriptor.enteringAnimation?.hashCode() ?: 0)
    }
    return hash
  }

  private fun spatialSubsample(
    markers: List<MarkerDescriptor>,
    maxCount: Int,
    bounds: LatLngBounds,
  ): Array<MarkerDescriptor> {
    val columns = ceil(sqrt(maxCount.toDouble())).toInt()
    val rows = ceil(maxCount.toDouble() / columns).toInt()

    val latMin = bounds.southwest.latitude
    val latMax = bounds.northeast.latitude
    val lonMin = bounds.southwest.longitude
    val lonMax = bounds.northeast.longitude

    val latStep = max(1e-9, (latMax - latMin) / rows)
    val lonStep = max(1e-9, (lonMax - lonMin) / columns)

    val buckets = LinkedHashMap<String, MutableList<MarkerDescriptor>>()

    for (marker in markers) {
      val row = minOf(rows - 1, maxOf(0, ((marker.coordinate.latitude - latMin) / latStep).toInt()))
      val column = minOf(columns - 1, maxOf(0, ((marker.coordinate.longitude - lonMin) / lonStep).toInt()))
      val key = "$row-$column"
      buckets.getOrPut(key) { mutableListOf() }.add(marker)
    }

    return buckets.values.map { cell -> cell[cell.size / 2] }.toTypedArray()
  }

  private fun maxMarkersForZoom(latitudeSpan: Double): Int {
    return when {
      latitudeSpan < 0.08 -> 2_000
      latitudeSpan < 0.5 -> 800
      latitudeSpan < 2.0 -> 350
      else -> 200
    }
  }

  private fun LatLngBounds.expandBy(fraction: Double): LatLngBounds {
    val latSpan = northeast.latitude - southwest.latitude
    val lngSpan = northeast.longitude - southwest.longitude
    val latPad = latSpan * fraction
    val lngPad = lngSpan * fraction

    return LatLngBounds(
      LatLng(southwest.latitude - latPad, southwest.longitude - lngPad),
      LatLng(northeast.latitude + latPad, northeast.longitude + lngPad),
    )
  }
}
